### PR TITLE
Refactored flusher logic

### PR DESCRIPTION
**Pull request by cpheps**
## Description of Changes
- Refactored flusher to not keep a context so it can use one passed in by the operator
- Added some retry configuration options
- Did some general cleanup of flusher code and added more unit test coverage.

## **Please check that the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
~~- [ ] Add a changelog entry (for non-trivial bug fixes / features)~~ (Part of larger v2 changelog)
- [x] CI passes

### CODE REVIEW

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.7586143|+0.18963516|53.839306|-74.71443|
|1|5000|5.75871|+0.5690799|57.386585|-80.06897|
|1|10000|12.862109|+1.827508|64.42969|-81.69841|
|1|50000|54.000668|+0.6900482|86.12769|-99.493935|
|1|100000|101.86217|+0.1343689|153.30711|-85.19409|
|10|100|2.6207063|+0.7241427|53.237206|-79.5167|
|10|500|7.138073|+1.3104548|58.713364|-84.810745|
|10|1000|14.103754|+1.1378384|63.301857|-84.6343|
|10|5000|60.07122|+3.0018196|97.16406|-83.43373|
|10|10000|118.44657|+7.4084167|165.6809|-72.77843|

File: docs/types/flusher.md [hunk 1]
```diff
@@ -1,10 +1,12 @@
 # Flushers
 
-Flushers handle reading entries from buffers in chunks, flushing them to their final destination, and retrying on failure.
+Flushers handles flushing entries to their final destination, and retrying on failure.
 
 In most cases, the default options will work well, but they may be need tuning for optimal performance or for reducing load
 on the destination API.
 
+The flusher can be used alongside a with the [buffer](./buffer.md). Tweaking configurations on both can help optimize sending entries to a destination.
```
Comment (BinaryFissionGames) [inline docs/types/flusher.md:8]:
```suggestion
The flusher can be used alongside a [buffer](./buffer.md). Tweaking configurations on both can help optimize sending entries to a destination.
```

File: docs/types/flusher.md [hunk 2]
```diff
@@ -19,3 +21,5 @@ Flushers are configured with the `flusher` block on output plugins.
 | Field               | Default | Description                                                                                                                                   |
 | ---                 | ---     | ---                                                                                                                                           |
 | `max_concurrent`    | `16`    | The maximum number of goroutines flushing entries concurrently                                                                                |
+| `max_retry_time`    | `1h`    | The maximum amount of time to retry flushing for                                                                                |
+| `max_retry_interval`    | `1m`    | Retry attempts us exponential backoffs. This value is the maximum interval reached during backoff retry.                                                                             |
```
Comment (BinaryFissionGames) [inline docs/types/flusher.md:25]:
```suggestion
| `max_retry_interval`    | `1m`    | Retry attempts use exponential backoffs. This value is the maximum interval reached during backoff retry.                                                                             |
```

File: operator/flusher/flusher.go [hunk 3]
```diff
@@ -3,136 +3,160 @@ package flusher
...
 	backoff "github.com/cenkalti/backoff/v4"
 	"go.uber.org/zap"
 	"golang.org/x/sync/semaphore"
 )
 
-// These are vars so they can be overridden in tests
-var maxRetryInterval = time.Minute
-var maxElapsedTime = time.Hour
+const (
+	// maxConcurrency is the default maximum amount of concurrent flush operatorations
```
Comment (BinaryFissionGames) [inline operator/flusher/flusher.go:14]:
```suggestion
	// maxConcurrency is the default maximum amount of concurrent flush operations
```

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.6034533|+0.034474134|51.31573|-77.23801|
|1|5000|5.6897483|+0.50011826|56.204338|-81.25121|
|1|10000|9.965736|-1.0688648|62.52896|-83.59914|
|1|50000|47.207317|-6.103302|91.32489|-94.29674|
|1|100000|82.71193|-19.01587|768.49786|+529.99664|
|10|100|2.6552267|+0.75866306|51.334724|-81.41918|
|10|500|6.5345063|+0.7068882|60.787178|-82.73693|
|10|1000|13.276089|+0.31017303|64.61786|-83.3183|
|10|5000|57.56958|+0.5001793|105.24273|-75.355064|
|10|10000|107.59285|-3.4453049|150.01172|-88.4476|

File: docs/types/flusher.md [hunk 4]
```diff
@@ -1,10 +1,12 @@
 # Flushers
 
-Flushers handle reading entries from buffers in chunks, flushing them to their final destination, and retrying on failure.
+Flushers handles flushing entries to their final destination, and retrying on failure.
```
Comment (jmwilliams89) [inline docs/types/flusher.md:3]: Should probably be `handle` singular here.

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.5517828|-0.017196298|52.260643|-76.2931|
|1|5000|6.3966765|+1.2070465|57.361126|-80.09442|
|1|10000|10.620811|-0.41378975|61.017242|-85.110855|
|1|50000|52.31036|-1.0002594|91.59523|-94.0264|
|1|100000|87.37957|-14.348228|137.44801|-101.05319|
|10|100|2.5172112|+0.62064755|54.00175|-78.75215|
|10|500|8.138084|+2.3104663|59.672413|-83.8517|
|10|1000|13.586316|+0.6204004|66.22441|-81.71175|
|10|5000|58.79414|+1.7247391|95.78394|-84.81385|
|10|10000|112.21586|+1.1777039|146.58217|-91.87715|

**Review by BinaryFissionGames — APPROVED** [review]

**Review by jmwilliams89 — APPROVED** [review]

**Pull request merged**